### ISSUE TITLE
IPC::Open3 - Fails if STDOUT is not opened to fileno 1 in caller

### ISSUE THREAD

**Issue by dfskoll** [Perl/perl5#19573]
If STDOUT is not open with fileno 1, then IPC::Open3 can fail.  For example, if STDOUT is opened for in-memory I/O, then its fileno is -1 and the file descriptors in the child process are not duplicated properly.  The attached test program illustrates the issue; expected output is:
```
Without messing with STDOUT:
Received: Hello
With STDOUT changed to in-memory I/O:
Received: Hello
```

but actual output is:

```
Without messing with STDOUT:
Received: Hello
With STDOUT changed to in-memory I/O:
Received: NOTHING RECEIVED
```

Open3 should make sure in the child that STDIN/STDOUT/STDERR are open with fileno 0, 1 and 2 before it dups the file descriptors.
[open3-bug-test_pl.txt](https://github.com/Perl/perl5/files/8383789/open3-bug-test_pl.txt)

**Comment by dfskoll** [comment]
The attached patch seems to fix it, but I haven't tested extensively.  It's patched against the version shipped with Debian 11; the patch applies right after the `untie` calls in the child process.
[open3-patch.txt](https://github.com/Perl/perl5/files/8383915/open3-patch.txt)

**Comment by dfskoll** [comment]
This patch (against the version of `IPC::Open3` shipped with Debian 11) seems to fix it.  Haven't done extensive testing.
[open3-patch.txt](https://github.com/Perl/perl5/files/8383943/open3-patch.txt)

**Comment by Grinnz** [comment]
As I alluded to in the PR, most things cannot be expected to function correctly if the program has closed STDIN, STDOUT, or STDERR, and then opened filehandles in their place (in-memory or otherwise). Programs should instead reopen the global handles to `File::Spec->devnull` if they are attempting to daemonize.

**Comment by dfskoll** [comment]
I disagree.  There are many use-cases for opening STDOUT to an in-memory string.  In our specific case, it's a Plack service that is accumulating its result in memory before sending it to the browser.

The problem is that IPC-Open3 conflates STDOUT with file descriptor 1.  Once we remove that assumption, it can easily be made to work without affecting existing code (see https://github.com/Perl/perl5/pull/19582) So I see absolutely no downside to fixing it.

**Comment by Grinnz** [comment]
I would recommend looking at how [Capture::Tiny](https://metacpan.org/pod/Capture::Tiny) accomplishes such things without breaking forks.

**Comment by dfskoll** [comment]
Hi,

On Mon, 04 Apr 2022 08:52:15 -0700
Dan Book ***@***.***> wrote:

> I would recommend looking at how
> [Capture::Tiny](https://metacpan.org/pod/Capture::Tiny) accomplishes
> such things without breaking forks.

Could you clarify what you mean by "breaking forks"?

Regards,

Dianne.

**Comment by Grinnz** [comment]
Well more specifically, breaking any code that uses STDOUT/STDIN/STDERR afterward, including forked processes.

**Comment by dfskoll** [comment]
My patch (the new one at https://github.com/Perl/perl5/pull/19582) does not break that.

In the case where we exec, STDIN/STDOUT/STDERR are irrelevant; they vaporize and the new program only knows about file descriptors 0, 1 and 2.  If the newly-execed program happens to be perl, it will set up STDIN/STDOUT/STDERR as per usual.

In the case where we do not exec, I do open STDIN/STDOUT to file descriptors 0 and 1 respectively (in the child... the parent is not affected), because otherwise the pipes will not work as expected.  If the child process doesn't want that behavior, then it can readjust STDIN and STDOUT to its taste.  But in my opinion, the behavior of open3 should be such that the child process can count on STDIN and STDOUT correctly referring to the pipes set up by the parent.

Regards,

Dianne.

**Comment by dfskoll** [comment]
On Mon, 04 Apr 2022 09:03:27 -0700
Dan Book ***@***.***> wrote:

> Well more specifically, breaking any code that uses
> STDOUT/STDIN/STDERR afterward, including forked processes.

My patch does not break that.

In the case where we exec, STDIN/STDOUT/STDERR are irrelevant; they
vaporize and the new program only knows about file descriptors 0, 1
and 2.  If the newly-execed program happens to be perl, it will set up
STDIN/STDOUT/STDERR as per usual.

In the case where we do not exec, I do open STDIN/STDOUT to file
descriptors 0 and 1 respectively (in the child... the parent is not
affected), because otherwise the pipes will not work as expected.  If
the child process doesn't want that behavior, then it can readjust
STDIN and STDOUT to its taste.

Regards,

Dianne.
